src/transport/MessageEndpointServer.cpp

```diff
@@ -0,0 +1,108 @@
+#include <faabric/transport/MessageEndpointServer.h>
+
+// Defined in libzmq/include/zmq.h (156384765)
+#define ZMQ_ETERM ETERM
+
+namespace faabric::transport {
+MessageEndpointServer::MessageEndpointServer(const std::string& hostIn,
+                                             int portIn)
+  : host(hostIn)
+  , port(portIn)
+{}
+
+void MessageEndpointServer::start()
+{
+    start(faabric::transport::getGlobalMessageContext());
+}
+
+void MessageEndpointServer::start(faabric::transport::MessageContext& context)
+{
+    // Start serving thread in background
+    this->servingThread = std::thread([this, &context] {
+        MessageEndpointClient serverEndpoint(this->host, this->port);
+
+        // Open message endpoint, and bind
+        serverEndpoint.open(
+          context, faabric::transport::SocketType::PULL, true);
+
+        // Loop until context is terminated
+        while (true) {
+            try {
+                this->recv(serverEndpoint);
+            } catch (zmq::error_t& e) {
+                if (e.num() == ZMQ_ETERM) {
+                    serverEndpoint.close();
+                    break;
+                }
+                throw std::runtime_error(fmt::format(
+                  "Errror in socket receiving message: {}", e.what()));
+            }
+        }
+    });
+}
+
+void MessageEndpointServer::stop()
+{
+    stop(faabric::transport::getGlobalMessageContext());
+}
+
+void MessageEndpointServer::stop(faabric::transport::MessageContext& context)
+{
+    // Note - different servers will concurrently close the server context, but
+    // this structure is thread-safe, and the close operation idempotent.
+    context.close();
+
+    // Finally join the serving thread
+    if (this->servingThread.joinable()) {
+        this->servingThread.join();
+    }
+}
+
+void MessageEndpointServer::recv(MessageEndpointClient& endpoint)
+{
+    // Receive header
+    if (!endpoint.socket) {
+        throw std::runtime_error("Trying to recv from a null-pointing socket");
+    }
+
+    zmq::message_t header;
+    if (!endpoint.socket->recv(header)) {
+        throw std::runtime_error("Error receiving message through socket");
+    }
+
+    // Check the header was sent with ZMQ_SNDMORE flag
+    if (!header.more()) {
+        throw std::runtime_error("Header sent without SNDMORE flag");
+    }
+
+    // Receive body
+    zmq::message_t body;
+    if (!endpoint.socket->recv(body)) {
```

Comment:
Probably getting into personal preference here but unless `recv` returns a `bool` I would be more explicit in the condition check, e.g. `if(endpoint.socket->recv(body) <= 0)` as it's easier to read, easier to diagnose when something goes wrong, and in this case I assume means number of bytes received?.

I would also try and include more information in these error messages as they're a bit vague, for example:

```
int bytesReceived = endpoint.socket->recv(body);
if(bytesReceived <= 0) {
    throw std::runtime_error(fmt::format("Error receiving message through socket, got {} bytes", bytesReceived));
}
```

Reply:
After checking the sources, it is a bit more involved than that. In particular, [`zmq::recv`](https://github.com/zeromq/cppzmq/blob/b65dde8725b2eb6d174a10e9e57891de554fd175/zmq.hpp#L1960) returns an optional value: either `numBytes` if it goes right, or an empty object `{}` otherwise. The C++ binding (cppzmq, the link I share before), already does the actual checks. IMO this is fine like it is.